pkg/processor/procbuiltin/unwraprecord.go

```diff
@@ -0,0 +1,185 @@
+		data := getSetter.Get(rec) // record.payload.after
+		var jsonData record.StructuredData
+		switch d := data.(type) {
+		case record.RawData:
+			// unmarshal raw data to structured
+			err = json.Unmarshal(data.Bytes(), &jsonData)
+			if err != nil {
+				return record.Record{}, cerrors.Errorf("%s: %w", processorType, err)
+			}
+		case record.StructuredData:
+			jsonData = d
+		default:
+			return record.Record{}, cerrors.Errorf("%s: unexpected data type %T", processorType, data)
+		}
+
+		payloadJSON, ok := jsonData[DebeziumFieldPayload]
+		if !ok {
+			return record.Record{}, cerrors.Errorf("%s: the payload is missing from the record", processorType)
+		}
+
+		r.Position = rec.Position
+		r.Key = rec.Key
+
+		// cast payload into a map
+		payload := payloadJSON.(map[string]any)
+
+		if format == RecordFormatKafkaConnect {
+			r.Payload.After = convertJSONToStructuredData(payload)
+			r.Operation = record.OperationSnapshot
+			return r, nil
+		}
+
+		// validate all the debezium fields exist
+		err = checkDebeziumFieldsExist(payload)
+		if err != nil {
+			return record.Record{}, cerrors.Errorf("%s: %w", processorType, err)
+		}
+
+		// convert json data into a record
+		before := convertJSONToStructuredData(payload[DebeziumFieldBefore])
+		after := convertJSONToStructuredData(payload[DebeziumFieldAfter])
+		metadata := convertJSONToMetadata(payload[DebeziumFieldSource])
+
+		r.Payload.Before = before
+		r.Payload.After = after
+		r.Metadata = metadata
+		r.Operation = getRecordOp(payload[DebeziumFieldOp].(string))
+
+		return r, nil
+	}), nil
+}
+
+const (
+	DebeziumOpCreate = "c"
+	DebeziumOpUpdate = "u"
+	DebeziumOpDelete = "d"
+	DebeziumOpRead   = "r" // snapshot
+
+	DebeziumFieldBefore  = "before"
+	DebeziumFieldAfter   = "after"
+	DebeziumFieldSource  = "source"
+	DebeziumFieldOp      = "op"
+	DebeziumFieldPayload = "payload"
+)
+
+func getRecordOp(o string) record.Operation {
+	switch o {
+	case DebeziumOpCreate:
+		return record.OperationCreate
+	case DebeziumOpUpdate:
+		return record.OperationUpdate
+	case DebeziumOpDelete:
+		return record.OperationDelete
+	case DebeziumOpRead:
+		return record.OperationSnapshot
+	}
+	return 0 // invalid op
+}
+
+func checkDebeziumFieldsExist(data record.StructuredData) error {
+	var multiErr error
+	if _, ok := data[DebeziumFieldBefore]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldBefore))
+	}
+	if _, ok := data[DebeziumFieldAfter]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldAfter))
+	}
+	if _, ok := data[DebeziumFieldSource]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldSource))
+	}
+	if _, ok := data[DebeziumFieldOp]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldOp))
+	}
```

Comment:
The op field may not be present. A Debezium connector can also include schema changes, in which case the op field is not present. As for `before` and `after`, there's virtually no difference between them being `null` and not being present (especially because some JSON serialization frameworks will simply skip them if they're `null`). With that, IMHO, returning errors in these cases is a bit too strict. 

I'm not sure about the source field. It looks like Debezium's source code says that `source` is optional too: https://github.com/debezium/debezium/blob/8ca5e0a9b51e5a14ac9e8fd342ffae3e12052428/debezium-core/src/main/java/io/debezium/data/Envelope.java#L243-L251

Reply:
we're unwrapping the record that was created here https://github.com/ConduitIO/conduit-connector-sdk/blob/6cd5101ba73f7388dfea47dc188692d27845d455/record.go#L252-L259 , so all these fields should be there.
I'm not sure if this is too specific or if we should make it more generic tbh@samirketema @lovromazgon what do you think? is it possible that platform will have a record that doesn't contain these fields? if so, should we still support it?

Reply:
The unwrap should work in the opposite way as the record formatter you linked to @maha-hajja, but it's not necessarily always going to be a record that comes from Conduit, it could be a record that is actually produced by Debezium.

That said, I thought we are focusing on unwrapping CDC records, as in [create](https://debezium.io/documentation/reference/2.1/connectors/postgresql.html#postgresql-create-events), [update](https://debezium.io/documentation/reference/2.1/connectors/postgresql.html#postgresql-update-events) and [delete](https://debezium.io/documentation/reference/2.1/connectors/postgresql.html#postgresql-delete-events) events. What's left are [truncate](https://debezium.io/documentation/reference/2.1/connectors/postgresql.html#postgresql-truncate-events), [message](https://debezium.io/documentation/reference/2.1/connectors/postgresql.html#postgresql-message-events) and [transaction](https://debezium.io/documentation/reference/2.1/connectors/postgresql.html#postgresql-transaction-metadata) events, but those would be much harder to parse into OpenCDC, we would need to first think about how we want to properly do that. Let's see what @samirketema says and if we can live without support for those events.

Reply:
> it could be a record that is actually produced by Debezium.

I was about to say the same thing. A "real" Debezium connector may have created this record and we got it like that. If this processor is about unwrapping Debezium records, then it should rely on the Debezium record "specification".

> Let's see what @samirketema says and if we can live without support for those events.

IIUC, we actually have 2 options. One is to let the Debezium un-wrapper drop the "non-row" events and another is to let a user do that (maybe via a built-in processor?)

Reply:
I have a bad feeling about an unwrap processor that is dropping records 😕 Right now if we got such a record the pipeline would stop because of an error. This feels safer to me compared to dropping the record and assuming the user doesn't care about it. I'd rather annoy the user and be loud about the fact that we don't support that record than drop it.

Reply:
Same here, I believe simply dropping records would be a surprise (unless we make it completely explicit through the processor name and the docs that the processor does so). IMHO it would be nice to have a processor which keeps only "row records" and drops "non-row" records.

Reply:
Ok then let's leave it like it is and see how far we get with this solution.

---

Regarding the processor that drops "non-row" records - this is doable right now by using two processors, first `parsejsonpayload` and then `filterfieldpayload` which can be configured to drop records where the field `op` doesn't contain `c`, `d`, `u` or `r`. I assume you are talking about creating a dedicated processor that does this out of the box - we need to figure out what our philosophy regarding that will be. I'm worried that if we start creating dedicated processors for things like this we risk creating a long list of processors and confusing our users about which one to use.

That's a conversation for another time and place, let's get this one shipped now 🙂

Reply:
I agree with leaving things as they are when it comes to "non-row" records. But IMHO we still shouldn't validate Debezium records in the way we do it now (existence of before, after, op fields).

Reply:
According to the Debezium specs I linked in a previous comment these 4 fields will be present in all records that the processor targets right now, why would we relax the condition?

Reply:
You're right about the `op` field. But the `before` and `after` fields are optional. For example:

![Screenshot from 2023-01-31 17-22-18](https://user-images.githubusercontent.com/2166300/215819337-bfb79e0f-9a1e-45ae-a87a-67871a44c581.png)

Reply:
Maybe we understand "optional" differently 😄 I thought it means that the field can be `null`, not that it might be missing entirely. For instance in a "create" event the field `before` will always be set to `null` since there was nothing there before, but they don't remove the field, they just set it to `null`. I'd expect the same behaviour for all change data events.

Reply:
FTR we discussed at the standup. In addition to the comment above, we may want to discuss the validation a bit more (how to make sure this processor is actually called on a Debezium record? what if we have unknown fields?), but the team agrees none of this is a blocker.